cmd/zfs/zfs_main.c

```diff
@@ -4383,6 +4383,100 @@ zfs_do_snapshot(int argc, char **argv)
 	return (-1);
 }
 
+typedef struct zfs_send_exclude_arg {
+	size_t count;
+	char **list;
+} zfs_send_exclude_arg_t;
+
+/*
+ * This function creates the zfs_send_exclude_arg_t
+ * object described above; it can be called multiple
+ * times, and the input can be comma-separated.
+ * This is NOT the most efficient data layout; however,
+ * I couldn't think of a non-pathological case where
+ * it should have more than a couple dozen instances
+ * of excludes. If that turns out to be used in
+ * practice, we might want to instead use a tree.
+ */
+static void
+add_dataset_excludes(char *in_exclude, zfs_send_exclude_arg_t *context)
+{
+	char *exclude = strdup(in_exclude);
+	size_t tempSize = context->count;
+	char **tempArray = context->list;
+	char *tempString = NULL;
+	char *exclude_orig = exclude;
+
+	while ((tempString = strsep(&exclude, ",")) != NULL) {
+		tempArray = realloc(tempArray,
+		    (sizeof (char **)) * (tempSize + 1));
+		if (tempArray == NULL)
+			nomem();
+		tempArray[tempSize] = strdup(tempString);
+		if (tempArray[tempSize] == NULL)
+			nomem();
+		tempSize++;
+	}
+
+	free(exclude_orig);
+	context->count = tempSize;
+	context->list = tempArray;
+}
+
+static void
+free_dataset_excludes(zfs_send_exclude_arg_t *exclude_list)
+{
+	if (exclude_list->count > 0) {
+		for (size_t i = 0; i < exclude_list->count; i++)
+			free(exclude_list->list[i]);
+		free(exclude_list->list);
+	}
+}
+
+/*
+ * This is the call back used by zfs_send to
+ * determine if a dataset should be skipped.
+ * As stated above, this is not the most efficient
+ * data structure to use, but as long as the
+ * number of excluded datasets is relatively
+ * small (a couple of dozen or so), it won't
+ * have a big impact on performance on modern
+ * processors. Since it's excluding hierarchies,
+ * we'd probably want to move to a more complex
+ * tree structure in that case.
+ */
+static boolean_t
+zfs_do_send_exclude(zfs_handle_t *zhp, void *context)
+{
+	zfs_send_exclude_arg_t *exclude = context;
+	const char *poolname = zfs_get_pool_name(zhp);
+	const char *fullname = zfs_get_name(zhp);
+	size_t indx;
+	const char *name;
+
+	name = fullname + strlen(poolname);
```

Comment:
is this just a really round-about way of doing `const char *name = strchrnul(fullname, '/');`?

Reply:
Quite likely, as that was a new function to me.  I was a bit cautious about this because I did NOT have the pool name in some of my tests cases! (And I still have no idea how that came about.)